src/LndClient/Data/LndEnv.hs

```diff
@@ -84,30 +91,40 @@ newLndEnv ::
   LndHexMacaroon ->
   LndHost ->
   LndPort ->
-  Either String LndEnv
+  Either LndError LndEnv
 newLndEnv pwd cert mac host port =
-  --
-  -- TODO : implement smart constructors
-  -- verify certificate and all other data validity
-  -- host, port etc
-  --
-  Right $
-    LndEnv
-      { envLndWalletPassword = pwd,
-        envLndHexMacaroon = mac,
-        envLndLogErrors = True,
-        envLndGrpcConfig =
-          ClientConfig
-            { clientServerHost = Host $ coerce host,
-              clientServerPort = Port $ coerce port,
-              clientArgs = [],
-              clientSSLConfig =
-                Just $
-                  ClientSSLConfig
-                    { serverRootCert = Just $ coerce cert,
-                      clientSSLKeyCertPair = Nothing,
-                      clientMetadataPlugin = Nothing
-                    },
-              clientAuthority = Nothing
-            }
-      }
+  createLndEnv <$> (validatePort port) <*> (validateCert cert)
+  where
+    validatePort :: LndPort -> Either LndError Int
+    validatePort p = do
+      let port32 :: Word32 = coerce p
+      let maybePort :: Maybe Int = U.safeFromIntegral port32
+      maybeToRight (LndEnvError "Wrong port") maybePort
+    validateCert :: LndTlsCert -> Either LndError LndTlsCert
+    validateCert c = do
+      pemsM <- mapLeft (\err -> LndEnvError $ LT.pack err) $ Pem.pemParseBS $ coerce c
```

Comment:
probably `mapLeft` might be replaced with `second` function?
because Either is instance of Bifunctor
maybe then we can remove `extra` dependency if it is not used in other places

Reply:
Btw, `(\err -> LndEnvError $ LT.pack err)` can be eta-reduced to `LndEnvError . pack`
https://wiki.haskell.org/Eta_conversion

Reply:
done, replaced to first
and did eta-convertion